internal/adaptorsigs/adaptor.go

```diff
@@ -0,0 +1,509 @@
+// a party who knows the hidden value, and a public key tweak can be done by
+// a party that only knows the point on the secp256k1 curve derived by the
+// multiplying the hidden value by the generator point.
+//
+// Generally the workflow of using adaptor signatures is the following:
+//  1. Party A randomly selects a hidden value and creates a private key
+//     modified adaptor signature of something for which party B requires
+//     a valid signature.
+//  2. The Party B sees the PublicTweak in the adaptor signature, and creates
+//     a public key tweaked adaptor signature for something that party A
+//     requires a valid signature.
+//  3. Since party A knows the hidden value, they can use the hidden value to
+//     create a valid signature from the public key tweaked adaptor signature.
+//  4. When the valid signature is revealed, by being posted to the blockchain,
+//     party B can recover the tweak and use it to decrypt the private key
+//     tweaked adaptor signature that party A originally sent them.
+type AdaptorSignature struct {
+	r           secp256k1.FieldVal
+	s           secp256k1.ModNScalar
+	t           affinePoint
+	pubKeyTweak bool
+}
+
+// Serialize returns a serialized adaptor signature in the following format:
+//
+//	sig[0:32]  x coordinate of the point R, encoded as a big-endian uint256
+//	sig[32:64] s, encoded also as big-endian uint256
+//	sig[64:96] x coordinate of the point T, encoded as a big-endian uint256
+//	sig[96] first bit is 1 if the signature is public key tweaked, second bit
+//	        is 1 if the y coordinate of T is odd.
+func (sig *AdaptorSignature) Serialize() []byte {
+	var b [AdaptorSignatureSize]byte
+	sig.r.PutBytesUnchecked(b[0:32])
+	sig.s.PutBytesUnchecked(b[32:64])
+	sig.t.x.PutBytesUnchecked(b[64:96])
+	if sig.pubKeyTweak {
+		b[96] = 1
+	}
+	b[96] |= byte(sig.t.y.IsOddBit()) << 1
+	return b[:]
+}
+
+// ParseAdaptorSignature parses an adaptor signature from a serialized format.
+func ParseAdaptorSignature(b []byte) (*AdaptorSignature, error) {
+	if len(b) != AdaptorSignatureSize {
+		str := fmt.Sprintf("malformed signature: wrong size: %d", len(b))
+		return nil, errors.New(str)
+	}
+
+	var r secp256k1.FieldVal
+	if overflow := r.SetBytes((*[32]byte)(b[0:32])); overflow > 0 {
+		str := "invalid signature: r >= field prime"
+		return nil, errors.New(str)
+	}
+
+	var s secp256k1.ModNScalar
+	if overflow := s.SetBytes((*[32]byte)(b[32:64])); overflow > 0 {
+		str := "invalid signature: s >= group order"
+		return nil, errors.New(str)
+	}
+
+	var t affinePoint
+	if overflow := t.x.SetBytes((*[32]byte)(b[64:96])); overflow > 0 {
+		str := "invalid signature: t.x >= field prime"
+		return nil, errors.New(str)
+	}
+
+	isOdd := (b[96]>>1)&1 == 1
+	if valid := secp256k1.DecompressY(&t.x, isOdd, &t.y); !valid {
+		str := "invalid signature: not for a valid curve point"
+		return nil, errors.New(str)
+	}
+	t.y.Normalize()
```

Comment:
This will no longer be necessary after https://github.com/decred/dcrd/pull/3441 lands.